File: app/src/main/java/com/example/reminderapp/AddReminderFragment.java
package com.example.reminderapp;

import android.app.AlarmManager;
import android.app.DatePickerDialog;
import android.app.PendingIntent;
import android.app.TimePickerDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.reminderapp.database.ReminderEntity;
import com.example.reminderapp.databinding.FragmentAddReminderBinding;

import java.util.Calendar;
import java.util.Locale;

public class AddReminderFragment extends Fragment {

    private FragmentAddReminderBinding binding;
    private ReminderViewModel viewModel;
    private final Calendar calendar = Calendar.getInstance();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentAddReminderBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        viewModel = new ViewModelProvider(this).get(ReminderViewModel.class);

        setupTypeSpinner();
        setupDateTimePickers();

        binding.btnSave.setOnClickListener(v -> saveReminder());
    }

    private void setupTypeSpinner() {
        String[] types = {"Normal", "Call Someone"};
        ArrayAdapter<String> adapter = new ArrayAdapter<>(requireContext(), android.R.layout.simple_spinner_item, types);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.spType.setAdapter(adapter);

        binding.spType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                binding.callFields.setVisibility(position == 1 ? View.VISIBLE : View.GONE);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void setupDateTimePickers() {
        binding.etDate.setOnClickListener(v -> {
            new DatePickerDialog(requireContext(), (view, y, m, d) -> {
                calendar.set(Calendar.YEAR, y);
                calendar.set(Calendar.MONTH, m);
                calendar.set(Calendar.DAY_OF_MONTH, d);
                binding.etDate.setText(String.format(Locale.getDefault(), "%d-%d-%d", y, m + 1, d));
            }, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH)).show();
        });

        binding.etTime.setOnClickListener(v -> {
            new TimePickerDialog(requireContext(), (view, h, min) -> {
                calendar.set(Calendar.HOUR_OF_DAY, h);
                calendar.set(Calendar.MINUTE, min);
                calendar.set(Calendar.SECOND, 0);
                binding.etTime.setText(String.format(Locale.getDefault(), "%02d:%02d", h, min));
            }, calendar.get(Calendar.HOUR_OF_DAY), calendar.get(Calendar.MINUTE), true).show();
        });
    }

    private void saveReminder() {
        String title = binding.etTitle.getText().toString();
        String type = binding.spType.getSelectedItem().toString();
        String contactName = binding.etContactName.getText().toString();
        String phoneNumber = binding.etPhoneNumber.getText().toString();
        String date = binding.etDate.getText().toString();
        String time = binding.etTime.getText().toString();
        String notes = binding.etNotes.getText().toString();

        if (title.isEmpty() || date.isEmpty() || time.isEmpty()) {
            Toast.makeText(requireContext(), "Please fill title, date and time", Toast.LENGTH_SHORT).show();
            return;
        }

        if (type.equals("Call Someone") && (contactName.isEmpty() || phoneNumber.isEmpty())) {
            Toast.makeText(requireContext(), "Please fill contact details", Toast.LENGTH_SHORT).show();
            return;
        }

        SharedPreferences prefs = requireContext().getSharedPreferences("user_prefs", Context.MODE_PRIVATE);
        int userId = prefs.getInt("active_user_id", -1);

        ReminderEntity reminder = new ReminderEntity(
            userId, title, notes, date, time, "", "None", "Personal", false, type, contactName, phoneNumber, calendar.getTimeInMillis(), false
        );

        viewModel.insert(reminder, id -> {
            reminder.setId((int) id);
            scheduleAlarm(reminder);
            Toast.makeText(requireContext(), "Reminder saved", Toast.LENGTH_SHORT).show();
            getParentFragmentManager().popBackStack();
        });
    }

    private void scheduleAlarm(ReminderEntity reminder) {
        AlarmManager alarmManager = (AlarmManager) requireContext().getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(requireContext(), CallReminderReceiver.class);
        intent.putExtra("id", reminder.getId());
        intent.putExtra("name", reminder.getContactName());
        intent.putExtra("phone", reminder.getPhoneNumber());
        intent.putExtra("notes", reminder.getNotes());
        intent.putExtra("type", reminder.getType());
        
        PendingIntent pendingIntent = PendingIntent.getBroadcast(
            requireContext(), 
            reminder.getId(), 
            intent, 
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
        );

        if (alarmManager != null) {
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, reminder.getDateTime(), pendingIntent);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/reminderapp/AddReminderFragment.kt
package com.example.reminderapp

import android.app.AlarmManager
import android.app.DatePickerDialog
import android.app.PendingIntent
import android.app.TimePickerDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.example.reminderapp.database.Reminder
import com.example.reminderapp.databinding.FragmentAddReminderBinding
import java.util.*

class AddReminderFragment : Fragment() {

    private var _binding: FragmentAddReminderBinding? = null
    private val binding get() = _binding!!
    private val viewModel: ReminderViewModel by viewModels()
    private val calendar = Calendar.getInstance()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentAddReminderBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupTypeSpinner()
        setupDateTimePickers()

        binding.btnSave.setOnClickListener {
            saveReminder()
        }
    }

    private fun setupTypeSpinner() {
        val types = arrayOf("Normal", "Call Someone")
        val adapter = ArrayAdapter(requireContext(), android.R.layout.simple_spinner_item, types)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spType.adapter = adapter

        binding.spType.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                binding.callFields.visibility = if (position == 1) View.VISIBLE else View.GONE
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun setupDateTimePickers() {
        binding.etDate.setOnClickListener {
            DatePickerDialog(requireContext(), { _, y, m, d ->
                calendar.set(Calendar.YEAR, y)
                calendar.set(Calendar.MONTH, m)
                calendar.set(Calendar.DAY_OF_MONTH, d)
                binding.etDate.setText("$y-${m + 1}-$d")
            }, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH)).show()
        }

        binding.etTime.setOnClickListener {
            TimePickerDialog(requireContext(), { _, h, min ->
                calendar.set(Calendar.HOUR_OF_DAY, h)
                calendar.set(Calendar.MINUTE, min)
                calendar.set(Calendar.SECOND, 0)
                binding.etTime.setText(String.format("%02d:%02d", h, min))
            }, calendar.get(Calendar.HOUR_OF_DAY), calendar.get(Calendar.MINUTE), true).show()
        }
    }

    private fun saveReminder() {
        val title = binding.etTitle.text.toString()
        val type = binding.spType.selectedItem.toString()
        val contactName = binding.etContactName.text.toString()
        val phoneNumber = binding.etPhoneNumber.text.toString()
        val date = binding.etDate.text.toString()
        val time = binding.etTime.text.toString()
        val notes = binding.etNotes.text.toString()

        if (title.isBlank() || date.isBlank() || time.isBlank()) {
            Toast.makeText(requireContext(), "Please fill title, date and time", Toast.LENGTH_SHORT).show()
            return
        }

        if (type == "Call Someone" && (contactName.isBlank() || phoneNumber.isBlank())) {
            Toast.makeText(requireContext(), "Please fill contact details", Toast.LENGTH_SHORT).show()
            return
        }

        val userId = requireContext().getSharedPreferences("user_prefs", Context.MODE_PRIVATE).getInt("active_user_id", -1)

        val reminder = Reminder(
            userId = userId,
            title = title,
            type = type,
            contactName = contactName,
            phoneNumber = phoneNumber,
            date = date,
            time = time,
            dateTime = calendar.timeInMillis,
            notes = notes
        )

        viewModel.insert(reminder) { id ->
            scheduleAlarm(reminder.copy(id = id.toInt()))
            Toast.makeText(requireContext(), "Reminder saved", Toast.LENGTH_SHORT).show()
            parentFragmentManager.popBackStack()
        }
    }

    private fun scheduleAlarm(reminder: Reminder) {
        val alarmManager = requireContext().getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(requireContext(), CallReminderReceiver::class.java).apply {
            putExtra("id", reminder.id)
            putExtra("name", reminder.contactName)
            putExtra("phone", reminder.phoneNumber)
            putExtra("notes", reminder.notes)
            putExtra("type", reminder.type)
        }
        
        val pendingIntent = PendingIntent.getBroadcast(
            requireContext(), 
            reminder.id, 
            intent, 
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, reminder.dateTime, pendingIntent)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}